dohproxy/server_protocol.py

```diff
@@ -73,11 +73,25 @@ async def query_udp(self, dnsq, clientip, timeout=DEFAULT_TIMEOUT):
     async def query_tcp(self, dnsq, clientip, timeout=DEFAULT_TIMEOUT):
         qid = dnsq.id
         fut = asyncio.Future()
-        transport, _ = await self.loop.create_connection(
-            lambda: DNSClientProtocolTCP(dnsq, fut, clientip, logger=self.logger),
-            self.upstream_resolver,
-            self.upstream_port,
-        )
+        try:
+            transport, _ = await asyncio.wait_for(
+                self.loop.create_connection(
+                    lambda: DNSClientProtocolTCP(
+                        dnsq, fut, clientip, logger=self.logger
+                    ),
+                    self.upstream_resolver,
+                    self.upstream_port,
+                ),
+                timeout,
+            )
+        except asyncio.TimeoutError:
+            self.logger.debug(
+                "Timeout connecting to upstream resolver {}:{}".format(
+                    self.upstream_resolver, self.upstream_port
+                )
+            )
+            return None
+
         return await self._try_query(fut, qid, timeout, transport)
```

Comment:
To make sure the overall request is bounded to "timeout", it could be better to change this call to something similar to:
```
return await self._try_query(fut, qid, timeout - (end_time - start_time), transport)
```